File: android/app/src/main/java/com/dash/phonecon/MessageType.kt
package com.dash.phonecon

object MessageType {
    // Phone → Mac
    const val CALL_RINGING = "CALL_RINGING"
    const val CALL_ACTIVE = "CALL_ACTIVE"
    const val CALL_ENDED = "CALL_ENDED"
    const val PING = "PING"

    // Mac → Phone
    const val ANSWER = "ANSWER"
    const val REJECT = "REJECT"
    const val HANGUP = "HANGUP"
    const val PONG = "PONG"
    const val DIAL = "DIAL"
    const val MUTE = "MUTE"

    // Contacts CRUD
    const val REQUEST_CONTACTS = "REQUEST_CONTACTS"
    const val CONTACTS_RESULT = "CONTACTS_RESULT"
    const val CONTACT_ADD = "CONTACT_ADD"
    const val CONTACT_UPDATE = "CONTACT_UPDATE"
    const val CONTACT_DELETE = "CONTACT_DELETE"
    const val CONTACT_OP_RESULT = "CONTACT_OP_RESULT"

    // Call log
    const val REQUEST_CALL_LOG = "REQUEST_CALL_LOG"
    const val CALL_LOG_RESULT = "CALL_LOG_RESULT"

    // JSON field keys
    const val FIELD_TYPE = "type"
    const val FIELD_NUMBER = "number"
    const val FIELD_NAME = "name"
    const val FIELD_CONTACTS = "contacts"
    const val FIELD_CONTACT_ID = "contact_id"
    const val FIELD_SUCCESS = "success"
    const val FIELD_ERROR = "error"
    const val FIELD_CALLS = "calls"
    const val FIELD_CALL_TYPE = "call_type"
    const val FIELD_TIMESTAMP = "timestamp"
    const val FIELD_DURATION = "duration"
    const val FIELD_MUTED = "muted"
}

File: linux/src/dashphone/protocol/message_type.py
"""The wire protocol shared by the Android phone and this app.

This file must stay in sync with:
  - android/app/src/main/java/com/dash/phonecon/MessageType.kt
  - macos/DashPhone/Models/MessageType.swift

Every message is a single-line JSON object with at least a "type" field,
sent over the WebSocket connection on port 8765. Keep this file as the
single source of truth for the string values used on the Ubuntu side.
"""

from __future__ import annotations

from enum import Enum


class MessageType(str, Enum):
    """Inheriting from str lets us drop straight into JSON (e.g. json.dumps({"type": MessageType.ANSWER}))."""

    # Phone -> this app
    CALL_RINGING = "CALL_RINGING"
    CALL_ACTIVE = "CALL_ACTIVE"
    CALL_ENDED = "CALL_ENDED"
    PING = "PING"

    # This app -> phone
    ANSWER = "ANSWER"
    REJECT = "REJECT"
    HANGUP = "HANGUP"
    PONG = "PONG"
    DIAL = "DIAL"  # this app -> phone: place an outgoing call to FIELD_NUMBER via the phone's TelecomManager
    MUTE = "MUTE"  # this app -> phone: FIELD_MUTED true/false - mute/unmute the phone's mic during an active call

    # Contacts CRUD - bidirectional, always initiated by this app, phone replies
    REQUEST_CONTACTS = "REQUEST_CONTACTS"  # this app -> phone: "send me your full contact list"
    CONTACTS_RESULT = "CONTACTS_RESULT"  # phone -> this app: FIELD_CONTACTS = [{id, name, number}, ...]
    CONTACT_ADD = "CONTACT_ADD"  # this app -> phone: create a contact from FIELD_NAME/FIELD_NUMBER
    CONTACT_UPDATE = "CONTACT_UPDATE"  # this app -> phone: FIELD_CONTACT_ID + new FIELD_NAME/FIELD_NUMBER
    CONTACT_DELETE = "CONTACT_DELETE"  # this app -> phone: delete FIELD_CONTACT_ID
    CONTACT_OP_RESULT = "CONTACT_OP_RESULT"  # phone -> this app: FIELD_SUCCESS + optional FIELD_ERROR

    # Call log - this app -> phone request, phone -> this app reply
    REQUEST_CALL_LOG = "REQUEST_CALL_LOG"  # this app -> phone: "send me recent call history"
    CALL_LOG_RESULT = "CALL_LOG_RESULT"  # phone -> this app: FIELD_CALLS = [{number, name, type, time, duration}, ...]


# JSON field keys (kept as plain constants, matching FIELD_* in MessageType.kt)
FIELD_TYPE = "type"
FIELD_NUMBER = "number"
FIELD_NAME = "name"
FIELD_CONTACTS = "contacts"
FIELD_CONTACT_ID = "contact_id"
FIELD_SUCCESS = "success"
FIELD_ERROR = "error"
FIELD_CALLS = "calls"
FIELD_CALL_TYPE = "call_type"
FIELD_TIMESTAMP = "timestamp"
FIELD_DURATION = "duration"
FIELD_MUTED = "muted"


def parse_message_type(raw: dict) -> MessageType | None:
    """Return the MessageType for a decoded JSON message, or None if unknown/missing.

    Returning None (instead of raising) lets callers ignore messages from a
    future protocol version without crashing - the same "default: break"
    behaviour used on the Mac and Android sides.
    """
    raw_type = raw.get(FIELD_TYPE)
    try:
        return MessageType(raw_type)
    except ValueError:
        return None
